src/code/z_demo.c

```diff
@@ -1,172 +1,176 @@
+s16 gDungeonBossWarpSceneId;
 
-void Cutscene_Init(PlayState* play, CutsceneContext* csCtx) {
+void Cutscene_InitContext(PlayState* play, CutsceneContext* csCtx) {
     s32 i;
 
-    csCtx->state = CS_STATE_0;
-    csCtx->frames = 0;
-    csCtx->unk_0C = 0.0f;
-    play->csCtx.sceneCsCount = 0;
-    play->csCtx.currentCsIndex = 0;
+    csCtx->state = CS_STATE_IDLE;
+    csCtx->curFrame = 0;
+    csCtx->timer = 0.0f;
+    play->csCtx.scriptListCount = 0;
+    play->csCtx.scriptIndex = 0;
 
-    for (i = 0; i < ARRAY_COUNT(D_801F4DC8); i++) {
-        D_801F4DC8[i] = 0;
+    for (i = 0; i < ARRAY_COUNT(sCueTypeList); i++) {
+        sCueTypeList[i] = 0;
     }
 
-    D_801F4DE0 = 0;
+    gDisablePlayerCsModeStartPos = 0;
 
     Audio_SetCutsceneFlag(false);
 }
 
-void Cutscene_Start(PlayState* play, CutsceneContext* csCtx) {
-    csCtx->state = CS_STATE_1;
-    csCtx->playerAction = NULL;
+void Cutscene_StartManual(PlayState* play, CutsceneContext* csCtx) {
+    csCtx->state = CS_STATE_START;
+    csCtx->playerCue = NULL;
 }
 
-void Cutscene_End(PlayState* play, CutsceneContext* csCtx) {
-    if (csCtx->state != CS_STATE_4) {
-        csCtx->state = CS_STATE_3;
+void Cutscene_StopManual(PlayState* play, CutsceneContext* csCtx) {
+    if (csCtx->state != CS_STATE_RUN_UNSTOPPABLE) {
+        csCtx->state = CS_STATE_STOP;
     }
 }
 
-typedef void (*CutsceneStateHandler)(PlayState* play, CutsceneContext* csCtx);
+typedef void (*CutsceneHandler)(PlayState* play, CutsceneContext* csCtx);
 
-CutsceneStateHandler sCsStateHandlers1[] = {
-    Cutscene_DoNothing, // CS_STATE_0
-    func_800EA258,      // CS_STATE_1
-    Cutscene_DoNothing, // CS_STATE_2
-    func_800ED9C4,      // CS_STATE_3
-    Cutscene_DoNothing, // CS_STATE_4
+CutsceneHandler sManualCutsceneHandlers[] = {
+    CutsceneHandler_DoNothing,   // CS_STATE_IDLE
+    CutsceneHandler_StartManual, // CS_STATE_START
+    CutsceneHandler_DoNothing,   // CS_STATE_RUN
+    CutsceneHandler_StopManual,  // CS_STATE_STOP
+    CutsceneHandler_DoNothing,   // CS_STATE_RUN_UNSTOPPABLE
 };
 
-void Cutscene_Update1(PlayState* play, CutsceneContext* csCtx) {
-    if (gSaveContext.save.cutscene < 0xFFF0) {
-        sCsStateHandlers1[csCtx->state](play, csCtx);
+void Cutscene_UpdateManual(PlayState* play, CutsceneContext* csCtx) {
+    if (gSaveContext.save.cutsceneIndex < 0xFFF0) {
+        sManualCutsceneHandlers[csCtx->state](play, csCtx);
     }
 }
 
-CutsceneStateHandler sCsStateHandlers2[] = {
-    Cutscene_DoNothing, // CS_STATE_0
-    func_800EA2B8,      // CS_STATE_1
-    func_800ED980,      // CS_STATE_2
-    func_800EDA04,      // CS_STATE_3
-    func_800ED980,      // CS_STATE_4
+CutsceneHandler sScriptedCutsceneHandlers[] = {
+    CutsceneHandler_DoNothing,   // CS_STATE_IDLE
+    CutsceneHandler_StartScript, // CS_STATE_START
+    CutsceneHandler_RunScript,   // CS_STATE_RUN
+    CutsceneHandler_StopScript,  // CS_STATE_STOP
+    CutsceneHandler_RunScript,   // CS_STATE_RUN_UNSTOPPABLE
```

Comment:
Are the comments here just indicating which state the handler runs in?

Reply:
Yeah exactly, each function is run to the corresponding state